change-validation/src/main/scala/com/normation/plugins/changevalidation/NotificationService.scala

```diff
@@ -113,11 +113,29 @@ class EmailNotificationService {
   }
 }
 
-class NotificationService(
+trait NotificationService {
+  def sendNotification(step: WorkflowNode, cr: ChangeRequest): IOResult[Unit]
+}
+
+object NotificationService {
+  def apply(
+      emailService:   EmailNotificationService,
+      linkUtil:       LinkUtil,
+      configMailPath: String
+  ): NotificationService = {
+    new NotificationServiceImpl(
+      emailService,
```

Comment:
I think I'm against that idea - but I'm not sure to understand the goal. Is it a kind of default implementation ? 
It looks like it adds a lot of burden in maintenance (each change in constructor have to been done at one more place) and I fail to see the earning. Can you explain ?

Reply:
As I wrote a new interface `NotificationService` to be able to use a mock version of it in tests, I renamed the class to `NotificationServiceImpl`, so previously `new NoficationService(args...)` would be now just `NotificationService(args...)`. It was either `NotificationService(args...)` everywhere, or `new NotificationServiceImpl(args...)`. I prefered the first version as it was returning an actual `NotificationService`, the trait and not the implementation.  

It also can be seen as a constructor that hides the implementation, but now the impl remains public (because the actual implementation is tested, including methods not in the interface...) so it kinda defeats its own purpose because of that :sweat_smile:

Reply:
Resolved in https://github.com/Normation/rudder-plugins/pull/618/commits/e3e6696e56805494aa10d21a8202cf95284744bf !